src/encryption.rs

```diff
@@ -1,53 +1,72 @@
+use crate::ConfigureError;
 use log::debug;
 use sodiumoxide::base64::Variant;
 use sodiumoxide::base64::{decode, encode};
 use sodiumoxide::crypto::secretbox;
+use std::fmt;
 use std::fs::{read, write};
-use std::io::{Error, ErrorKind};
 use std::path::Path;
 
 pub fn init() {
     sodiumoxide::init().expect("Unable to initialize libsodium");
 }
 
-pub fn generate_key() -> String {
+pub fn generate_key() -> EncryptionKey {
     debug!("Generating an encryption key");
     let key_bytes = secretbox::gen_key();
-    encode_key(key_bytes)
+    EncryptionKey::from_str(&encode_key(&key_bytes)).expect("Unable to generate new encryption key")
 }
 
 pub fn encrypt_file(
     input_path: &Path,
     output_path: &Path,
-    secret: &str,
-) -> Result<(), std::io::Error> {
-    let content = read(input_path)?;
-    let ciphertext = encrypt_bytes(content, decode_key(secret));
-    write(&output_path, &ciphertext)?;
-
-    Ok(())
+    key: &EncryptionKey,
+) -> Result<(), ConfigureError> {
+    let file_contents = match read(input_path) {
+        Ok(file_contents) => file_contents,
+        Err(_err) => return Err(ConfigureError::InputFileNotReadable),
+    };
+
+    let encrypted_bytes = match encrypt_bytes(&file_contents, key) {
+        Ok(encrypted_bytes) => encrypted_bytes,
+        Err(_err) => return Err(ConfigureError::DataEncryptionError),
+    };
+
+    match write(&output_path, encrypted_bytes) {
+        Ok(()) => Ok(()),
+        Err(_err) => Err(ConfigureError::OutputFileNotWritable),
+    }
 }
 
 pub fn decrypt_file(
     input_path: &Path,
     output_path: &Path,
-    secret: &str,
-) -> Result<(), std::io::Error> {
-    let content = read(input_path)?;
-
-    match decrypt_bytes(content, decode_key(secret)) {
-        Ok(decrypted_bytes) => Ok(write(&output_path, decrypted_bytes)?),
-        Err(_err) => Err(Error::new(ErrorKind::InvalidData, "Unable to decrypt file")),
+    key: &EncryptionKey,
+) -> Result<(), ConfigureError> {
+    let file_contents = match read(input_path) {
+        Ok(file_contents) => file_contents,
+        Err(_err) => return Err(ConfigureError::InputFileNotReadable),
+    };
+
+    let decrypted_bytes = match decrypt_bytes(&file_contents, key) {
+        Ok(decrypted_bytes) => decrypted_bytes,
+        Err(_err) => return Err(ConfigureError::DataDecryptionError),
+    };
+
+    match write(&output_path, decrypted_bytes) {
+        Ok(()) => Ok(()),
+        Err(_err) => Err(ConfigureError::OutputFileNotWritable),
     }
 }
 
-fn encrypt_bytes(input: Vec<u8>, key: sodiumoxide::crypto::secretbox::Key) -> Vec<u8> {
+fn encrypt_bytes(input: &[u8], key: &EncryptionKey) -> Result<Vec<u8>, ConfigureError> {
```

Comment:
It doesn't look like this function is returning an error. I've tried the following diff and it built without any errors:

```
diff --git a/src/encryption.rs b/src/encryption.rs
index 984c947..842520f 100644
--- a/src/encryption.rs
+++ b/src/encryption.rs
@@ -59,7 +59,9 @@ pub fn decrypt_file(
     }
 }
 
-fn encrypt_bytes(input: &[u8], key: &EncryptionKey) -> Result<Vec<u8>, ConfigureError> {
+enum AnotherError {}
+
+fn encrypt_bytes(input: &[u8], key: &EncryptionKey) -> Result<Vec<u8>, AnotherError> {
     let nonce = secretbox::gen_nonce();
     let secret_bytes = secretbox::seal(input, &nonce, &key.key);
 ```

Reply:
You're right! I've removed that path in c2258851049afa353440098f1f8a9c5cd60bd936, but kept the error around in order to avoid throwing off the other error codes.